lib/Dialect/Torch/Transforms/MaximizeValueSemantics.cpp

```diff
@@ -148,40 +150,41 @@ class AbstractlyInterpretCopyToNonValueTensorOpUsersWithinABlock
     // terminating at CopyToValueTensorOp's, possibly with intervening view-like
     // ops and overwrites. This also catches the special case of a
     // CopyToNonValueTensorOp that trivially feeds into CopyToValueTensorOp's.
-    SmallVector<Operation *> nonValueTensorUsers;
-    auto workList = llvm::to_vector(copy.result().getUsers());
+    DenseMap<Operation *, SmallVector<Value>> nonValueTensorsUsedByOp;
+    using OpOperandRefs = SmallVector<std::reference_wrapper<OpOperand>>;
+    OpOperandRefs workList(copy.result().getUses());
     while (!workList.empty()) {
-      Operation *op = workList.pop_back_val();
+      OpOperand &operand = workList.pop_back_val();
+      Operation *op = operand.getOwner();
       if (op->getBlock() != copy->getBlock()) {
         return rewriter.notifyMatchFailure(
             copy, "can only analyze within a single basic block");
       }
-      nonValueTensorUsers.push_back(op);
+      nonValueTensorsUsedByOp[op].push_back(
+          assertNonValueTensor(operand.get()));
 
       if (isViewLikeOp(op)) {
-        auto isTensor = [](const Value operand) {
-          return operand.getType().isa<BaseTensorType>();
-        };
-
-        // We currently only support view-like ops with one tensor input and one
-        // tensor output, meaning that the tensor use-def chains form a tree.
-        // This will not be the case for an op like `torch.aten.view_as`, so
-        // we will need to add a set to prune duplicate visitation.
-        if (llvm::count_if(op->getOperands(), isTensor) != 1 ||
-            llvm::count_if(op->getResults(), isTensor) != 1 ||
-            !isTensor(op->getOperand(0)) || !isTensor(op->getResult(0))) {
+        // We currently only support view-like ops with one tensor output.
+        if (op->getNumResults() == 0 ||
+            !op->getResult(0).getType().isa<BaseTensorType>()) {
           return rewriter.notifyMatchFailure(
-              copy, "unsupported: view-like ops must have one tensor input and "
-                    "one tensor output, and the tensor input/output must be "
-                    "the first operand/result");
+              copy, "unsupported: view-like ops must have one tensor output, "
+                    "and the tensor output must be the first result");
         }
 
-        llvm::append_range(workList, op->getResult(0).getUsers());
+        Value opResult = op->getResult(0);
+        // There are cases where a view-like op will be partially converted to
+        // value semantics, resulting in at least one of the inputs being a
+        // non-value tensor and the output being a value tensor. If this is the
+        // case then there is no need to look at the users of the result of the
+        // op.
+        if (opResult.getType().isa<NonValueTensorType>())
```

Comment:
There are two assumptions here:
1. the first operand is aliased with the single result
2. the `copyToNonValueTensor` of the first operand is processed before the second tensor operand.

For 1, I think there is currently no accurate way to tell which operand aliasing with which result from a torch op at this moment. The jit operator schema does provide this information so we can add that in the future. For now, we might want to add a comment about this.

For 2, We could add some guard.

Reply:
Done!